applications/FluidDynamicsApplication/custom_processes/calulate_levelset_consistent_nodal_gradient_process.cpp

```diff
@@ -0,0 +1,148 @@
+
+    // Iterate over the elements
+    #pragma omp parallel for firstprivate(DN_DX, N, J0, InvJ0, detJ0, pressures, distances, grad)
+    for(int i_elem = 0; i_elem < elements_number; ++i_elem) {
+
+        const auto it_elem = it_element_begin + i_elem;
+        auto& r_geometry = it_elem->GetGeometry();
+
+        // Current geometry information
+        const unsigned number_of_nodes = r_geometry.PointsNumber();
+
+        for(std::size_t i_node=0; i_node<number_of_nodes; ++i_node){
+            distances(i_node) = r_geometry[i_node].FastGetSolutionStepValue(DISTANCE);
+        }
+
+        unsigned int nneg=0, npos=0;
+        for(unsigned int i = 0; i < number_of_nodes; ++i)
+        {
+            if(distances(i) > 0) {
+                npos += 1;
+            } else {
+                nneg += 1;
+            }
+        }
+
+        if(nneg == 0 || npos == 0)
+        {
+            // The integration points
+            const auto& r_integration_method = r_geometry.GetDefaultIntegrationMethod();
+            const auto& r_integration_points = r_geometry.IntegrationPoints(r_integration_method);
+            const unsigned int number_of_integration_points = r_integration_points.size();
+
+            for(std::size_t i_node=0; i_node<number_of_nodes; ++i_node)
+                pressures[i_node] = r_geometry[i_node].FastGetSolutionStepValue(PRESSURE);
+
+            // The containers of the shape functions and the local gradients
+            const auto& rNcontainer = r_geometry.ShapeFunctionsValues(r_integration_method);
+            const auto& rDN_DeContainer = r_geometry.ShapeFunctionsLocalGradients(r_integration_method);
+
+            for ( IndexType point_number = 0; point_number < number_of_integration_points; ++point_number ) {
+                // Getting the shape functions
+                noalias(N) = row(rNcontainer, point_number);
+
+                // Getting the jacobians and local gradients
+                GeometryUtils::JacobianOnInitialConfiguration(r_geometry, r_integration_points[point_number], J0);
+                MathUtils<double>::GeneralizedInvertMatrix(J0, InvJ0, detJ0);
+                const auto& rDN_De = rDN_DeContainer[point_number];
+                GeometryUtils::ShapeFunctionsGradients(rDN_De, InvJ0, DN_DX);
+
+                noalias(grad) = prod(trans(DN_DX), pressures);
+                const double gauss_point_volume = r_integration_points[point_number].Weight() * detJ0;
+
+                for(unsigned int i_node=0; i_node<number_of_nodes; ++i_node) {
+                    array_1d<double, 3>& r_gradient = r_geometry[i_node].GetValue(PRESSURE_GRADIENT);
+                    for(unsigned int k=0; k<num_dim; ++k) {
+                        #pragma omp atomic
+                        r_gradient[k] += N[i_node] * gauss_point_volume*grad[k];
+                    }
+
+                    double& r_vol = r_geometry[i_node].GetValue(NODAL_AREA);
+
+                    #pragma omp atomic
```

Comment:
Use `AddAtomic` from the `AtomicUtilities` in here.

Reply:
Done.